include/openssl/engine.h

```diff
@@ -798,9 +798,17 @@ typedef int (*dynamic_bind_engine) (ENGINE *e, const char *id,
         OPENSSL_EXPORT \
         int bind_engine(ENGINE *e, const char *id, const dynamic_fns *fns) { \
             if (ENGINE_get_static_state() == fns->static_state) goto skip_cbs; \
+            if (getenv("OPENSSL_engine_libcrypto_override") == NULL) { \
+                fprintf(stderr, "ERROR: engine load aborted because a " \
+                        "different libcrypto was detected!\n" \
+                        "To override this check, please set " \
+                        "OPENSSL_engine_libcrypto_override=yes."); \
+                return 0; \
+            } \
             CRYPTO_set_mem_functions(fns->mem_fns.malloc_fn, \
                                      fns->mem_fns.realloc_fn, \
                                      fns->mem_fns.free_fn); \
+            OPENSSL_init_crypto(OPENSSL_INIT_NO_ATEXIT, NULL); \
         skip_cbs: \
```

Comment:
The more I look at it, the more I would replace this entire section with this:

``` C
            if (ENGINE_get_static_state() != fns->static_state) return 0; \
```

In other words, not allowing the application libcrypto and the engine libcrypto to differ, at all

Reply:
I would agree, but unfortunately the problem here is, that it sometimes works, and we even have test
cases where this "feature" is used, so it will probably not be easy to get there in one step.

Reply:
> The more I look at it, the more I would replace this entire section with this:
> 
> ```c
>             if (ENGINE_get_static_state() != fns->static_state) return 0; \
> ```
> 
> In other words, not allowing the application libcrypto and the engine libcrypto to differ, at all

Well, it can be done without breaking tests, if my PR is taken, but only for linux. I am not pushing my PR any longer. Providers are a great fix to this problem.